lib/nostrum/voice/audio.ex

```diff
@@ -0,0 +1,234 @@
+defmodule Nostrum.Voice.Audio do
+  @moduledoc false
+
+  require Logger
+
+  alias Nostrum.Error.VoiceError
+  alias Nostrum.Struct.VoiceState
+  alias Nostrum.Util
+  alias Nostrum.Voice
+  alias Porcelain.Process, as: Proc
+
+  @encryption_mode "xsalsa20_poly1305"
+  @samples_per_frame 960
+  @usec_per_frame 20_000
+  # How many consecutive packets to send before resting
+  @frames_per_burst 10
+
+  def encryption_mode, do: @encryption_mode
+
+  def rtp_header(%VoiceState{} = voice) do
+    <<
+      0x80::8,
+      0x78::8,
+      voice.rtp_sequence::16,
+      voice.rtp_timestamp::32,
+      voice.ssrc::32
+    >>
+  end
+
+  def encrypt_packet(%VoiceState{} = voice, data) do
+    header = rtp_header(voice)
+    # 12 byte header + 12 null bytes
+    nonce = header <> <<0::96>>
+    header <> Kcl.secretbox(data, nonce, voice.secret_key)
+  end
+
+  def open_udp do
+    {:ok, socket} =
+      :gen_udp.open(0, [
+        :binary,
+        {:active, false},
+        {:reuseaddr, true}
+      ])
+
+    socket
+  end
+
+  def init_player(voice) do
+    take_nap()
+    player_loop(voice)
+  end
+
+  def player_loop(voice, init? \\ true) do
+    t1 = Util.usec_now()
+    voice = try_send_data(voice, init?)
+    t2 = Util.usec_now()
+
+    take_nap(t2 - t1)
+
+    player_loop(voice, false)
+  end
+
+  def take_nap(diff \\ 0) do
+    ((@usec_per_frame * @frames_per_burst - diff) / 1000)
+    |> trunc()
+    |> max(0)
+    |> Process.sleep()
+  end
+
+  def try_send_data(%VoiceState{} = voice, init?) do
+    wait = if(init?, do: 20_000, else: 200)
```

Comment:
How does this work? If someone plays a short song (under 20s) will we not send the event until after the 20s has elapsed? Or is this for some other functionality entirely? Not sure what `stalling` means in this context!